Sources/SourceControl/GitRepository.swift

```diff
@@ -81,6 +81,20 @@ public struct GitRepositoryProvider: RepositoryProvider {
         // expected cost of iterative updates on a full clone is less than on a
         // shallow clone.
         precondition(!localFileSystem.exists(path))
+
+        // Provides better feedback when mistakingly using url: for a dependancy that
+        // is a local package. Still allows for using url with a local package that has
+        // also been initialized by git
+        if let validPath = try? AbsolutePath(validating: repository.url), localFileSystem.exists(validPath) {
+            do {
+                try self.callGit("-C", repository.url, "rev-parse", "--git-dir", repository: repository)
+            } catch let error as GitCloneError {
+                throw GitCloneError(repository: repository,
+                                    message: "Cannot clone from local directory \(repository.url)\nUse path: instead of url: for \(repository.url)",
```

Comment:
the function "fetch" is generic, but the error message here assume a very specific usage / call-site. would it be better to throw a "InvalidPath(reason)" (or similar) error here, and then in the call site construct the more specific error message?

Reply:
We are emitting from `WorkspaceRepositoryManagerDelegate.fetchingDidFinish` directly to the diagnostics engine, so that's the closest to the call site we could get right now. That's a little less generic, but even there we don't have much information about the specific use case.

Reply:
I see. Incidentally worked on some code refactoring for identity as perhaps we can do this there: https://github.com/apple/swift-package-manager/pull/3239/files#diff-27a3fa78e1b60a72806fdffb7bcb91661c1ec5e9c15d0c69d0c0c96d24942c44R311?

Reply:
Yeah, we should be able to move it into PackageDescription4Loader.swift.